clang_delta/EmptyStructToInt.cpp

```diff
@@ -351,9 +451,112 @@ bool EmptyStructToInt::isValidRecordDecl(const RecordDecl *RD)
   return true;
 }
 
+const RecordDecl *EmptyStructToInt::getBaseRecordDef(const Type *Ty)
+{
+  const ArrayType *ArrayTy = dyn_cast<ArrayType>(Ty);
+  if (ArrayTy) {
+    Ty = getArrayBaseElemType(ArrayTy);
+  }
+
+  if (!Ty->isStructureType())
+    return NULL;
+
+  const RecordType *RT = Ty->getAsStructureType();
+  return RT->getDecl()->getDefinition();
+}
+
+void EmptyStructToInt::getInitExprs(const Type *Ty,
+                                           const Expr *E,
+                                           const IndexVector *IdxVec,
+                                           ExprVector &InitExprs)
+{
+  const ArrayType *ArrayTy = dyn_cast<ArrayType>(Ty);
+  if (ArrayTy) {
+    const InitListExpr *ILE = dyn_cast<InitListExpr>(E);
+    TransAssert(ILE && "Invalid array initializer!");
+    unsigned int NumInits = ILE->getNumInits();
+    Ty = ArrayTy->getElementType().getTypePtr();
+
+    for (unsigned I = 0; I < NumInits; ++I) {
+      const Expr *Init = ILE->getInit(I);
+      getInitExprs(Ty, Init, IdxVec, InitExprs);
+    }
+    return;
+  }
+
+  const InitListExpr *ILE = dyn_cast<InitListExpr>(E);
+  if (!ILE)
+    return;
+
+  const RecordType *RT = NULL;
+  if (Ty->isUnionType()) {
+    RT = Ty->getAsUnionType();
+  }
+  else if (Ty->isStructureType()) {
+    RT = Ty->getAsStructureType();
+  }
+  else {
+    TransAssert(0 && "Bad RecordType!");
+  }
+
+  const RecordDecl *RD = RT->getDecl();
+
+  if(RD->getCanonicalDecl() == TheRecordDecl)
+  {
+    InitExprs.push_back(E);
+  }
+  else
+  {
+    for(IndexVector::const_iterator FI = IdxVec->begin(), FE = IdxVec->end(); FI != FE; ++FI)
+    {
+      const FieldDecl *FD = getFieldDeclByIdx(RD, (*FI));
+      TransAssert(FD && "NULL FieldDecl!");
+
+      Ty = FD->getType().getTypePtr();
+
+      const RecordDecl *BaseRD = getBaseRecordDef(Ty);
+
+      unsigned int InitListIdx;
+
+      if(BaseRD->isUnion())
+      {
+        InitListIdx = 0;
+      }
+      else
+      {
+        InitListIdx = (*FI);
+      }
+
+      if(InitListIdx >= ILE->getNumInits())
+      {
+        return;
+      }
+
+      getInitExprs(Ty, ILE->getInit(InitListIdx), RecordDeclToField[BaseRD], InitExprs);
+    }
+  }
+}
+
+const FieldDecl *EmptyStructToInt::getFieldDeclByIdx(
```

Comment:
It's only called by ```getInitExprs```. Since the caller is dead, perhaps we should remove this one, too.